File: solution/kotlin/week06-exercise2/src/main/kotlin/report/ReportGenerator.kt
package report

import arrow.core.*
import report.ReportError.NoDataToReport
import report.ReportError.UnsupportedReportType

class ReportGenerator(private val reportGenerators: Map<ReportType, CanGenerateReport>) {
    fun generateReport(reportType: ReportType, data: List<ReportData>): Either<ReportError, String> {
        if (data.isEmpty()) return NoDataToReport("No data provided for report generation.").left()

        return when (val generator = Option.fromNullable(reportGenerators[reportType])) {
            is Some -> generator.value.generateReport(data)
            is None -> UnsupportedReportType("Report type $reportType not supported.").left()
        }
    }
}

File: solution/c#/week06-legacy-code/ReportGenerator/ReportGenerator.cs
using LanguageExt;

namespace ReportGenerator
{
    public class ReportGenerator(Dictionary<ReportType, ICanGenerateReport> reportGenerators)
    {
        public Either<ReportError, string> GenerateReport(ReportType reportType, Seq<ReportData> data)
        {
            if (data.IsEmpty)
            {
                return new NoDataToReport("No data provided for report generation.");
            }

            return reportGenerators.TryGetValue(reportType, out var value)
                ? value.GenerateReport(data)
                : new UnsupportedReportType($"Report type {reportType} not supported.");
        }
    }
}